src/hazelcore/semantics/UHPat.re

```diff
@@ -153,6 +158,51 @@ and make_inconsistent_operand =
     (Parenthesized(set_p), u_gen);
   };
 
+let patterns_of_type =
+    (u_gen: MetaVarGen.t, ty: HTyp.t): (list(t), MetaVarGen.t) =>
+  OpSeq.(
+    switch (ty) {
+    | HTyp.Hole
+    | Unit
+    | Arrow(_, _) => ([], u_gen)
+    | Num => ([numlit(0), wild()] |> List.map(wrap), u_gen)
+    | Bool => ([true, false] |> List.map(b => wrap(boollit(b))), u_gen)
+    | Prod(_, _) =>
+      let prod_element_count = ty |> HTyp.get_prod_elements |> List.length;
+      let patternSkel =
+        ListUtil.range(prod_element_count)
+        |> List.map(i => Skel.Placeholder(i))
+        |> make_tuple(NotInHole);
+      let opseq_of_holes = u_gen => {
+        let (firstHole, u_gen) = u_gen |> new_EmptyHole;
+        let (holes, u_gen) = new_EmptyHoles(prod_element_count - 1, u_gen);
+        let opseq =
+          OpSeq(
+            patternSkel,
+            Seq.mk(firstHole, holes |> List.map(hole => (Comma, hole))),
+          );
+        (opseq, u_gen);
+      };
+
+      let (opseq1, u_gen) = opseq_of_holes(u_gen);
+      let (opseq2, u_gen) = opseq_of_holes(u_gen);
+      ([opseq1, opseq2], u_gen);
+    | Sum(_, _) =>
+      let (holes, u_gen) = new_EmptyHoles(2, u_gen);
+      let patterns =
+        List.map2(
+          (side, hole) => wrap(injection(side, wrap(hole))),
+          InjSide.[L, R],
+          holes,
+        );
+      (patterns, u_gen);
+    | List(_) =>
+      let (hole1, u_gen) = u_gen |> new_EmptyHole;
```

Comment:
can use new_EmptyHoles here too

Reply:
I could, by I would have to destructure the list in the assignment (or rather pattern match on a list with 2 elements). The compiler will complain about not having all cases handled, so I would have to disable the warning.
The current approach is less concise, but it's safe by construction, so overall I think it's better.

Reply:
ah fair